File: cpp/Problem55.cpp
/*
 * Problem: 55. Jump Game
 * URL: https://leetcode.com/problems/jump-game/
 * Difficulty: Medium
 * Tags: Array, Math
 */

#include <vector>

using namespace std;

class Solution {
public:
	bool canJump(vector<int>& nums) {
		int endIndex = nums.size() - 1;
		for (int i = endIndex; i >= 0; i--) {
			if (i + nums[i] >= endIndex) {
				endIndex = i;
			}
		}
		return endIndex == 0;
	}
};

File: python/Problem55.py
# Problem: 55. Jump Game
# URL: https://leetcode.com/problems/jump-game/
# Difficulty: Medium
# Tags: Array, Math

from typing import List

class Solution:
  def canJump(self, nums: List[int]) -> bool:
    endIndex: int = len(nums) - 1
    for i in range(endIndex, 0, -1):
      if i + nums[i] >= endIndex:
        endIndex = i
    return endIndex == 0